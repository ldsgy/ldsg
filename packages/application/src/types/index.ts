export * from "./handler-resource-settings";
