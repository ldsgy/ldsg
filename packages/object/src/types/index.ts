export * from "./get-field-info";
export * from "./specific-resource-settings";
