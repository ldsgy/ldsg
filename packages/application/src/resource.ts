import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
