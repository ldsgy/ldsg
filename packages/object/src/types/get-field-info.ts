import { PlatformParams } from "@ldsg/field-type";
import { GeneralResourceSettings } from "@ldsg/types";

export interface FieldInfo extends GeneralResourceSettings {
  /**
   * Field Name
   */
  name: string;

  /**
   * Type Info
   */
  typeInfo: any;
}

interface GetFieldInfoRes {
  fieldInfo: FieldInfo;
}

export type GetFieldInfo = (params: PlatformParams) => GetFieldInfoRes;
