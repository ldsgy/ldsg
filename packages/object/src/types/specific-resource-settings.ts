import { SpecificResourceSettings } from "@ldsg/types";

export interface ObjectSpecificResourceSettings
  extends SpecificResourceSettings {}
