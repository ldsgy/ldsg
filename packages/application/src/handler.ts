import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
