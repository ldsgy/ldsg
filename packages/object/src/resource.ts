import { PlatformsParams } from "@ldsg/field-type";
import { Resource } from "@ldsg/resource";
import { GeneralResourceSettings } from "@ldsg/types";
import _ from "lodash";
import {
  FieldInfo,
  GetFieldInfo,
  ObjectSpecificResourceSettings,
} from "./types";

export interface ExtendedObjectInfo {
  /**
   * Field Info List
   */
  fieldInfoList: FieldInfo[];
}

export interface GetExtendedObjectInfoRes {
  /**
   * Extended Object Info
   */
  extendedObjectInfo: ExtendedObjectInfo;
}

type GetExtendedObjectInfo = (
  params?: PlatformsParams
) => GetExtendedObjectInfoRes;

export interface ObjectInfo extends GeneralResourceSettings {
  /**
   * Object Name
   */
  name: string;

  /**
   * Field Info List
   */
  fieldInfoList: FieldInfo[];
}

export interface GetObjectInfoRes {
  /**
   * Object Info
   */
  objectInfo: ObjectInfo;
}

type GetObjectInfo = (params?: PlatformsParams) => GetObjectInfoRes;

export class ObjectResource extends Resource<ObjectSpecificResourceSettings> {
  getExtendedObjectInfo: GetExtendedObjectInfo = (params) => {
    const {
      settings: { extendedResourceIds },
      getFilteredResource,
    } = this;

    const fieldInfoListGroup = _.map(
      extendedResourceIds,
      (extendedResourceId) => {
        const { resource } = getFilteredResource<ObjectResource>({
          id: extendedResourceId,
        });

        const getObjectInfoRes = resource?.getObjectInfo(params);

        const res = getObjectInfoRes?.objectInfo.fieldInfoList || [];

        return res;
      }
    );

    const fieldInfoList = _.flatten(fieldInfoListGroup);

    const extendedObjectInfo: ExtendedObjectInfo = {
      fieldInfoList,
    };

    const res = {
      extendedObjectInfo,
    };

    return res;
  };

  getObjectInfo: GetObjectInfo = (params) => {
    const {
      id,
      settings: { title, description, name },
      getExtendedObjectInfo,
      getFilteredResources,
    } = this;

    const { extendedObjectInfo } = getExtendedObjectInfo(params);

    const { fieldInfoList: extendedFieldInfoList } = extendedObjectInfo;

    const { resources } = getFilteredResources<{ getFieldInfo?: GetFieldInfo }>(
      {
        parentId: id,
      }
    );

    const mapRes = resources.map((resource) => {
      const fieldInfo = resource.getFieldInfo?.(params);

      return fieldInfo;
    });

    const filterRes = _.filter(mapRes, (value) => !_.isUndefined(value));

    const currentFieldInfoList = _.map(filterRes, "fieldInfo");

    const fieldInfoList = extendedFieldInfoList.concat(currentFieldInfoList);

    const objectInfo: ObjectInfo = {
      title,
      description,
      name,
      fieldInfoList,
    };

    const res = {
      objectInfo,
    };

    return res;
  };
}
