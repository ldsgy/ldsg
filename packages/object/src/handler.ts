import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { ObjectResource } from "./resource";
import { ObjectSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<ObjectSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<ObjectResource>
> = instantiateResource;
