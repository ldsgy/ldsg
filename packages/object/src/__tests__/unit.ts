import { FieldTypeResource } from "@ldsg/field-type";
import { Handler, HandlerResource } from "@ldsg/handler";
import { Resource } from "@ldsg/resource";
import { SpecificResourceSettings } from "@ldsg/types";
import { ObjectResource } from "../resource";
import { FieldInfo, GetFieldInfo } from "../types";

interface ASpecificResourceSettings extends SpecificResourceSettings {
  /**
   * Field Name
   */
  name: string;

  /**
   * Field Type Resource ID
   */
  fieldTypeResourceId: string;

  /**
   * Field Properties
   */
  properties: any;
}

class AResource extends Resource<ASpecificResourceSettings> {
  getFieldInfo: GetFieldInfo = (params) => {
    const { platform } = params;

    const {
      settings: { title, description, name, properties },
      getResourcesFromSettings,
    } = this;

    const { fieldTypeResource } = getResourcesFromSettings();

    const typeInfo = (fieldTypeResource as FieldTypeResource).getFieldTypeInfo({
      platform,
      fieldProperties: properties,
    });

    const fieldInfo: FieldInfo = {
      title,
      description,
      name,
      typeInfo,
    };

    const res = {
      fieldInfo,
    };

    return res;
  };
}

const handler: Handler<
  [
    {
      /**
       * Field Properties
       */
      fieldProperties?: any;
      /**
       * Platform
       * Such as mongoose\formily.
       */
      platform: string;
    }
  ],
  {}
> = (params) => {
  const { fieldProperties, platform } = params;

  const { max } = fieldProperties;

  let res;

  switch (platform) {
    case "mongoose": {
      res = {
        type: "String",
        ...(max
          ? {
              maxLength: max,
            }
          : {}),
      };

      break;
    }

    default: {
      res = {};

      break;
    }
  }

  return res;
};

test("object", () => {
  new HandlerResource({
    id: "test-handler",
    kind: "HANDLER",
    parentId: "root",
    settings: {
      title: "测试处理程序",
      description: "",
      code: "",
      dependencies: [],
    },
    handler,
  });

  new FieldTypeResource({
    id: "test-field-type",
    kind: "FIELD_TYPE",
    parentId: "root",
    settings: {
      title: "文本",
      description: "可用来存储各种文本",
      fieldPropertiesSchema: {
        type: "object",
        properties: {
          max: {
            type: "integer",
            title: "配置最长字符",
            description: "长度不可以超过此值",
          },
        },
      },
      handlerResourceId: "test-handler",
    },
  });

  const objectResource = new ObjectResource({
    id: "test-object",
    kind: "OBJECT",
    parentId: "root",
    settings: {
      title: "测试对象",
      description: "",
      name: "test-object",
    },
  });

  new AResource({
    id: "test-a-1",
    kind: "OBJECT_FIELD",
    parentId: "test-object",
    settings: {
      title: "测试字段1",
      description: "",
      name: "test1",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new AResource({
    id: "test-a-2",
    kind: "OBJECT_FIELD",
    parentId: "test-object",
    settings: {
      title: "测试字段2",
      description: "",
      name: "test2",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new AResource({
    id: "test-a-3",
    kind: "OBJECT_FIELD",
    parentId: "test-object",
    settings: {
      title: "测试字段3",
      description: "",
      name: "test",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  const getObjectInfoRes = objectResource.getObjectInfo({
    platform: "mongoose",
  });

  expect(getObjectInfoRes).toMatchSnapshot();
});
