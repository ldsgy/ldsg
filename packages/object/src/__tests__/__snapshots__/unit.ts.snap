// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object 1`] = `
{
  "objectInfo": {
    "description": "",
    "fieldInfoList": [
      {
        "description": "",
        "name": "test1",
        "title": "测试字段1",
        "typeInfo": {
          "maxLength": "10",
          "type": "String",
        },
      },
      {
        "description": "",
        "name": "test2",
        "title": "测试字段2",
        "typeInfo": {
          "maxLength": "10",
          "type": "String",
        },
      },
      {
        "description": "",
        "name": "test",
        "title": "测试字段3",
        "typeInfo": {
          "maxLength": "10",
          "type": "String",
        },
      },
    ],
    "name": "test-object",
    "title": "测试对象",
  },
}
`;
